src/lib/ndpi_main.c

```diff
@@ -4536,12 +4536,16 @@ int load_categories_file_fd(struct ndpi_detection_module_struct *ndpi_str,
       category = strtok_r(NULL, "\t", &saveptr);
 
       if(category) {
-	int rc = ndpi_load_category(ndpi_str, name,
-				    (ndpi_protocol_category_t) atoi(category),
-				    user_data);
+        cat_id = atoi(category);
+        if(cat_id > 0 && cat_id < NDPI_PROTOCOL_NUM_CATEGORIES) {
+
+          int rc = ndpi_load_category(ndpi_str, name,
+				      (ndpi_protocol_category_t)cat_id,
+				      user_data);
```

Comment:
Hmmm... Personally, I'd replace such stuff with `strtonum`. Ofc, this is a BSD specific function, but I guess it's easy to write our own implementation from scratch.

@IvanNardi @utoni what do you think about it?

Reply:
I agree. I remember that I've tried to basically copy&paste the function from `libbsd` a long time ago, but gave up as it did not compile on MingW/MSVC. But if you find a proper solution, I'll approve. :)

Reply:
> I agree. I remember that I've tried to basically copy&paste the function from `libbsd` a long time ago, but gave up as it did not compile on MingW/MSVC. But if you find a proper solution, I'll approve. :)

Not giga concise, but it works. Requires only stdlib.h, limits.h and errno.h
```
int64_t ndpi_strtonum(const char* numstr, int64_t minval, int64_t maxval, const char** errstrp)
{
  int64_t retval = 0;
  char* endptr;

  if (minval > maxval) {
    *errstrp = "minval > maxval";
    return retval;
  }

  retval = (int64_t)strtoll(numstr, &endptr, 10);

  if (numstr == endptr || *endptr != '\0') {
    *errstrp = "invalid value";
     return 0;
  } else if ((ret == LLONG_MIN && errno == ERANGE) || (ret < minval)) {
    *errstrp = "value too small";
     return 0;
  } else if ((ret == LLONG_MAX && errno == ERANGE) || (ret > maxval ))
    *errstrp = "value too large";
     return 0;
  }

  return retval;
}
```

Reply:
Switched to `ndpi_strtonum`